### ISSUE TITLE
[Regression in PR][$250] Clicking on "New Messages" jumps to newest messages, skipping over any messages marked unread report by @chauchausoup

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#11584]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Go to any chat
2. Click at Mark as Unread
3. Click on New Messages button

## Expected Result:

Clicking "new messages" should jump directly to the message marked as Unread 

## Actual Result:

Clicking "new messages" jumps to the bottom of the message thread, skipping over the message marked as unread

## Workaround:
Unknown

## Platform:

Where is this issue occurring?

- Web

**Version Number:** 1.2.11-2
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**Email or phone of affected tester (no customers):** 
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 
https://user-images.githubusercontent.com/43996225/193866710-b8402785-759a-4f1f-aa86-80b0878cb3c9.mov
https://user-images.githubusercontent.com/43996225/193866742-10e455e5-acd7-4d2d-862a-b61a85c3e9c5.mp4
**Expensify/Expensify Issue URL:** 
**Issue reported by:** @chauchausoup
**Slack conversation:** https://expensify.slack.com/archives/C01GTK53T8Q/p1664725500869019
[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by greg-schroeder** [comment]
I reproduced this on the desktop app:

https://user-images.githubusercontent.com/29386373/193938022-1bc5a558-2b98-4196-84f0-9276480058bc.mp4

**Comment by conorpendergrast** [comment]
Oops, I'll get to this today, sorry!

**Comment by tienifr** [comment]
# Proposal
## Problem
- Currently, the logic scroll to unread message is using `scrollToBottom ` function
https://github.com/Expensify/App/blob/7b21103bb56eb0cf3fc1561715613d2a16c3255f/src/pages/home/report/ReportActionsView.js#L285

- function `ReportScrollManager.scrollToIndex` isn't work correctly in web because it has different behavior scroll from native

## Solution
In https://github.com/Expensify/App/blob/7b21103bb56eb0cf3fc1561715613d2a16c3255f/src/pages/home/report/ReportActionsView.js#L286
we'll change the function name `scrollToBottomAndMarkReportAsRead ` to `scrollToUnreadMsgAndMarkReportAsRead `
we'll also change to use `scrollToIndex`. In theory, we just need to pass this.props.report.lastReadSequenceNumber as the index but we're using inverted flatList so we have to pass `this.props.report.maxSequenceNumber - this.state.newMarkerSequenceNumber`


```diff
-        this.scrollToBottomAndMarkReportAsRead = this.scrollToBottomAndMarkReportAsRead.bind(this);
+       this.scrollToUnreadMsgAndMarkReportAsRead = this.scrollToUnreadMsgAndMarkReportAsRead.bind(this);
```


```diff
-    scrollToBottomAndMarkReportAsRead() {
-        ReportScrollManager.scrollToBottom();
+    scrollToUnreadMsgAndMarkReportAsRead() {
+        ReportScrollManager.scrollToIndex({animated: true, index: this.props.report.maxSequenceNumber - this.state.newMarkerSequenceNumber}, false);
        Report.readNewestAction(this.props.report.reportID);
    }
```

```diff
-                            onClick={this.scrollToBottomAndMarkReportAsRead}
+                           onClick={this.scrollToUnreadMsgAndMarkReportAsRead}

```
In https://github.com/Expensify/App/blob/main/src/components/InvertedFlatList/BaseInvertedFlatList.js
we need to measure again because the image can be loaded later. And we also pass sizeMap to FlatList to calculate the offset value

```diff
    measureItemLayout(nativeEvent, index) {
        const computedHeight = nativeEvent.layout.height;

 -       // We've already measured this item so we don't need to
 -       // measure it again.
 -       if (this.sizeMap[index]) {
 -           return;
 -       }
```


```diff
      <FlatList
                // eslint-disable-next-line react/jsx-props-no-spreading
                {...this.props}
                ref={this.props.innerRef}
                inverted
                renderItem={this.renderItem}
+              sizeMap={this.sizeMap}
```

In https://github.com/Expensify/App/blob/7b21103bb56eb0cf3fc1561715613d2a16c3255f/src/libs/ReportScrollManager/index.js#L14
we have to re-implement the logic scrollToIndex in non-native platform
because we're using inverted flatList so we have to calculate the total offset value from top to index, then + paddingTop, + viewOffset, - clientHeight * index.viewPosition if need. Finally pass to offset value in `scrollToOffset` function

```diff

-    flatListRef.current.scrollToIndex(index);
+    const sizeMap = flatListRef.current.props.sizeMap;
+    const clientHeight = flatListRef.current.getScrollableNode().clientHeight;
+    const paddingTop = parseFloat(flatListRef.current.getScrollableNode().childNodes[0].style.paddingTop || '0px');

+   const sizeArray = _.map(_.values(sizeMap), s => s.length);
+    const sumNum = _.reduce(sizeArray.slice(0, (index.index)), (acc, val) => acc + val);
+   flatListRef.current.scrollToOffset({
+       animated: index.animated,
+       offset: (sumNum + paddingTop + (index.viewOffset || 0)) - (clientHeight * (index.viewPosition || 0)),
+    });
}
```

**Comment by tienifr** [comment]
https://user-images.githubusercontent.com/113963320/195074818-5736b458-7c56-4b14-824a-ed1a6a7b533a.mp4

**Comment by tienifr** [comment]
https://user-images.githubusercontent.com/113963320/195074906-b0cd9614-8351-45fd-a4ea-808f74ec983b.mp4

**Comment by conorpendergrast** [comment]
Job posting: https://www.upwork.com/ab/applicants/1579975844227059712/job-details

**Comment by conorpendergrast** [comment]
@eVoloshchak Hey! Ready for your review on the proposal ^^^

**Comment by eVoloshchak** [comment]
Need to test this one thoroughly, expect a review in 24 hours

**Comment by tsa321** [comment]
Maybe  all unread indicator error should be on hold, because Roryabraham will implement new algorithm to solve it. Here is slack conversation:

https://expensify.slack.com/archives/C01GTK53T8Q/p1665794707786899

**Comment by eVoloshchak** [comment]
@tienifr proposal looks good, confirmed it working on all of the platforms
A couple of small things:

- Is there a way to obtain `clientHeight` and `paddingTop` differently? Maybe there are constants defined in `styles.js`? If not it's ok
- Could you please overexplain what is the `sizeArray` and `sumNum` variables and what we need them for?

**Comment by tienifr** [comment]
@eVoloshchak sorry, I don't catch on your mean
1. I'm updating the logic for `scrollToIndex` not `scrollToBottom`
2. We're implementing `invertedWheelEvent` in our project and FlatList in `react-native-web` works well,  so I don't think we should change the inverted lists inside of the package itself

**Comment by eVoloshchak** [comment]
> I'm updating the logic for scrollToIndex not scrollToBottom

I was talking about [this](https://github.com/Expensify/App/blob/7b21103bb56eb0cf3fc1561715613d2a16c3255f/src/libs/ReportScrollManager/index.js#L14) function, but yes, I can see that your implementation calls `scrollToBottom` inside of it

> We're implementing `invertedWheelEvent` in our project

Oh cool, didn't notice it. In that case your fix seems like the best solution. We essentially need to correctly calculate the offset to the first unread message and scroll to it, which it does. 

I assume you have already seen this but if not, please take a look at the [STYLE.md](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#accessing-object-properties-and-default-values) contributing guide before submitting the PR.

cc: @conorpendergrast, @tienifr's [proposal](https://github.com/Expensify/App/issues/11584#issuecomment-1274513698) looks good to me 
🎀👀🎀 C+ reviewed!

**Comment by conorpendergrast** [comment]
Great! @srikarparsi Are you happy here too? If so, I'll hire via Upwork ([this process](https://stackoverflow.com/c/expensify/questions/7676))

**Comment by srikarparsi** [comment]
yup this looks good to me as well, assigning @tienifr to the issue

**Comment by tienifr** [comment]
Thanks @srikarparsi , I applied to the job and the PR will be ready by October 21

**Comment by conorpendergrast** [comment]
Via Upwork, hired @tienifr to fix the issue, @chauchausoup for reporting the bug,  @eVoloshchak as contributor+

**Comment by conorpendergrast** [comment]
@srikarparsi @eVoloshchak @tienifr Hey! I've added [some proposed testing steps](https://github.com/Expensify/App/issues/11584#issuecomment-1294707771). Who can help with the second, third and fourth points (finding the PR that introduced the bug, the comments on that PR, and starting a discussion in #contributor-plus)? Then, I will get started on the last 3 steps!

**Comment by conorpendergrast** [comment]
Looks like we're all set to pay this. I'll review the process and will do that tomorrow!

**Comment by conorpendergrast** [comment]
And, also, [this checklist](https://github.com/Expensify/App/issues/11584#issuecomment-1294707771) above!

**Comment by conorpendergrast** [comment]
Fantastic, thanks @eVoloshchak !

**Comment by conorpendergrast** [comment]
Just writing down for myself that these steps are taken in order, so I'll be paying via UpWork ([this process](https://stackoverflow.com/c/expensify/questions/7582)) once steps 1 to 4 are done

**Comment by eVoloshchak** [comment]
The PR that introduced the bug has been identified. Link to the PR: https://github.com/Expensify/App/pull/4603 
`The issue exists from the creation of the component, probably just an edge case that wasn't tested (or  Mark as unread functionality didn't exist at the time)`

The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment: https://github.com/Expensify/App/pull/4603#issuecomment-1310192920

A discussion in #contributor-plus has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:  `Not sure about this, it's just an edge case that was missed`

cc: @conorpendergrast

**Comment by parasharrajat** [comment]
Not a regression but a new feature.

**Comment by conorpendergrast** [comment]
Ok, I've posted a recommended regression test ([internal channel link](https://expensify.slack.com/archives/C01SKUP7QR0/p1668154421465549)) and have posted the proposal to **not** add anything else to test for this ([#contributor-plus link](https://expensify.slack.com/archives/C02NK2DQWUX/p1668155104493619)). 

Last step now, to pay [the three people we hired](https://github.com/Expensify/App/issues/11584#issuecomment-1289713096).

**Comment by conorpendergrast** [comment]
All paid, and all set here. Thanks all!

**Comment by conorpendergrast** [comment]
Woops I missed [some steps](https://stackoverflow.com/c/expensify/questions/7582) last week; contracts ended and job removed now

**Comment by parasharrajat** [comment]
There is a new issue report which looks like a regression and should have been tested on this PR. https://expensify.slack.com/archives/C049HHMV9SM/p1668507952621799

**Comment by kavimuru** [comment]
@parasharrajat Reopening this issue as per this https://expensify.slack.com/archives/C049HHMV9SM/p1668537124299269?thread_ts=1668507952.621799&cid=C049HHMV9SM

**Comment by eVoloshchak** [comment]
@tienifr, could you take a look at [this](https://expensify.slack.com/archives/C049HHMV9SM/p1668507952621799) please?

**Comment by conorpendergrast** [comment]
@tienifr Bump for participating in [that regression discussion](https://expensify.slack.com/archives/C049HHMV9SM/p1668507952621799)!

**Comment by tienifr** [comment]
I'm looking at this issue @conorpendergrast @eVoloshchak

**Comment by tienifr** [comment]
@conorpendergrast sorry but i can't open your link, I think it's on private channel. Pls let me know if I miss something. Thanks

**Comment by eVoloshchak** [comment]
@tienifr, it's in `#expensify-bugs` channel, to get an invite just [send an email](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#slack-channels).

Meanwhile, I'll just copy the discussion from slack
Bug:  Clicking “New Message” doesn’t scroll to the correct chat.
Steps:

1. Go to a report that has a few 100 messages
2. Mark any message as unread
3. Scroll above the message you’ve marked as unread
4. Click the green “New message” button
5. Repeat step 4 multiple times

Expected: Screen scrolls to the message marked as unread
Actual: Screen keeps scrolling up
Version: v1.2.26-0 (production)
Platform: Web


https://user-images.githubusercontent.com/9059945/202998688-344eae05-a1ef-4da5-a2ba-e7d67edaa6a6.mov

**Comment by tienifr** [comment]
I've sent the email yesterday, still waiting to get added to the `#expensify-bugs` channel

**Comment by tienifr** [comment]
## Problem
- Currently when user deletes comment, in `reportActions_${id}` just set `html=''` in field `mesage` and do not change `maxSequenceNumber`. Please check evidence below:
 
![image](https://user-images.githubusercontent.com/113963320/203266573-7b740a23-ecdf-4b6c-a86a-34697bf8950f.png)

- So in https://github.com/Expensify/App/blob/7b21103bb56eb0cf3fc1561715613d2a16c3255f/src/pages/home/report/ReportActionsView.js#L284 formula for calculating index `index = this.props.report.maxSequenceNumber - this.state.newMarkerSequenceNumber` isn't work correctly in web because in list comment still has index of deleted comment.

## Solution
In https://github.com/Expensify/App/blob/7b21103bb56eb0cf3fc1561715613d2a16c3255f/src/pages/home/report/ReportActionsView.js#L284
we'll caculate `countDeletedAction` and recalculate the value of the `index` based on `countDeletedAction`.

```diff
  scrollToUnreadMsgAndMarkReportAsRead() {
+        let countDeletedAction = 0;
+        for (let i = this.state.newMarkerSequenceNumber; i < this.props.report.maxSequenceNumber; i++) {
+            if (ReportActionsUtils.isDeletedAction(this.props.reportActions[i])) {
+                countDeletedAction++;
+            }
+        }
-       ReportScrollManager.scrollToIndex({ animated: true, index: this.props.report.maxSequenceNumber - this.state.newMarkerSequenceNumber}, false);
+       ReportScrollManager.scrollToIndex({ animated: true, index: this.props.report.maxSequenceNumber - this.state.newMarkerSequenceNumber - countDeletedAction }, false);
       Report.readNewestAction(this.props.report.reportID);
    }
```
Before change:

https://user-images.githubusercontent.com/113963320/203266910-0aa04e4d-ec87-4ee5-b712-2d6b595e97cd.mp4


After change:


https://user-images.githubusercontent.com/113963320/203266971-9cb1824e-4a25-4c47-ac87-d6e304124893.mp4

**Comment by eVoloshchak** [comment]
@tienifr, nice job investigating this one!
I have a couple of questions
1. Can we actually change `maxSequenceNumber` to the correct one? Will this break something else?
2. Does this mean this bug shouldn't exist in chats that don't have deleted messages?

**Comment by tienifr** [comment]
1. I think when we change `maxSequenceNumber` to the correct one, scroll to new message will not work correctly.
Example: When we has 10 messages, delete the first 4 messages and we need to scroll to 9th message in `FlatList`
2. Yeah, i think so

**Comment by eVoloshchak** [comment]
Not overdue, will review the proposal tomorrow

**Comment by eVoloshchak** [comment]
@tienifr, this is looking much better, good job on discovering it!
There is still a bug however, if you mark a message unread and then delete a message from chat, the position to which it scrolls is wrong


https://user-images.githubusercontent.com/9059945/204526170-eff5a18d-a19e-4715-b6f1-d5f0c408d0b0.mov

**Comment by marcaaron** [comment]
Hey Everyone!

Clearly there is some confusion here caused by this issue here -> https://github.com/Expensify/App/issues/11584

From the look of it... it was not triaged correctly and should have been immediately closed as the "Expected behavior" listed is wrong. I'd suggest we revert things all the way back to how they were before [this PR](https://github.com/Expensify/App/pull/12124) was created and stop accepting proposals to fix anything related to this (and we should take this `Internal` for now to clean up the mess).

@srikarparsi can you make this a top priority? I think we should also [do a RCA](https://stackoverflow.com/c/expensify/questions/12156) here because a feature that should never have been implemented intentionally hit production. This is just my opinion, but to me that's _worse than any bug_. It signals that we're no longer in control and likely need to change something about our process.

**Comment by JmillsExpensify** [comment]
> This is just my opinion, but to me that's worse than any bug. It signals that we're no longer in control and likely need to change something about our process.

I agree with @marcaaron here. @conorpendergrast @srikarparsi I think we need to treat this an a top priority for today and [get the RCA published](https://stackoverflow.com/c/expensify/questions/12156). We've got to figure out how we avoid this in the future, and especially changing defined/designed behaviors in the app.

**Comment by conorpendergrast** [comment]
Before we do that, let's agree the behaviour we expect. Discussing that [in Slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1669844955940759?thread_ts=1669780795.674179&cid=C01GTK53T8Q) now!

**Comment by conorpendergrast** [comment]
Reversion PR is here: https://github.com/Expensify/App/pull/13230
And @srikarparsi is working on an RCA!

**Comment by conorpendergrast** [comment]
Scroll up, Melv

**Comment by srikarparsi** [comment]
Issue reverted here: https://github.com/Expensify/App/pull/13230

Will close after the PR gets deployed and I respond to @conorpendergrast's comments in the RCA

**Comment by srikarparsi** [comment]
PR to address one change in the RCA: https://github.com/Expensify/App/pull/13518
Will bring up the other proposed changes from the RCA in slack

**Comment by trjExpensify** [comment]
👋 Hey everyone! Can you give me an update on why this issue is still open, and what we need to do to close it out? I'm trying to piece it together:

1. The PR to revert the change has been merged
2. The linked PR with some action from the RCA has been closed #13518 
3. Potentially something about other proposed changed from the RCA are in slack, but not sure where that is or what (if any) actions we need to take from that? 

Thanks!

**Comment by srikarparsi** [comment]
hey @trjExpensify! I think this issue is ready to be closed out, the PR has been reverted, the RCA has been completed, and I'm talking to marc about the proposed changes in the RCA but I believe that can be continued with this issue closed. @tjferriss @tjferriss please reopen it though if there's something that I'm missing.
